contributors/design-proposals/cpu-manager.md

```diff
@@ -0,0 +1,348 @@
+  prev := m.shared
+  m.shared = c
+  for _, pod := range m.driver.GetPods() {
+    for _, container := range p.Containers {
+      if driver.GetCPUs(container.ID) == prev {
+        driver.SetCPUs(m.shared)
+      }
+    }
+  }
+}
+
+// @pre: container_qos = guaranteed
+func numExclusive(c v1.Container) int {
+  if c.resources.requests["cpu"] % 1000 == 0 {
+    return c.resources.requests["cpu"] / 1000
+  }
+  return 0
+}
+```
+
+##### Example pod specs and interpretation
+
+| Pod                                        | Interpretation                 |
+| ------------------------------------------ | ------------------------------ |
+| Pod [Guaranteed]:<br />&emsp;A:<br />&emsp;&emsp;cpu: 0.5 | Container **A** is assigned to the shared cpuset. |
+| Pod [Guaranteed]:<br />&emsp;A:<br />&emsp;&emsp;cpu: 2.0 | Container **A** is assigned two sibling threads on the same physical core (HT) or two physical cores on the same socket (no HT.)<br /><br /> The shared cpuset is shrunk to  make room for the exclusively allocated CPUs. |
+| Pod [Guaranteed]:<br />&emsp;A:<br />&emsp;&emsp;cpu: 1.0<br />&emsp;B:<br />&emsp;&emsp;cpu: 0.5 | Container **A** is assigned one exclusive CPU and container **B** is assigned to the shared cpuset. |
+| Pod [Guaranteed]:<br />&emsp;A:<br />&emsp;&emsp;cpu: 1.5<br />&emsp;B:<br />&emsp;&emsp;cpu: 0.5 | Both containers **A** and **B** are assigned to the shared cpuset. |
+| Pod [Burstable] | All containers are assigned to the shared cpuset. |
+| Pod [BestEffort] | All containers are assigned to the shared cpuset. |
+
+#### Policy 3: "dynamic" cpuset control
+
+_TODO: Describe the policy._
+
+##### Implementation sketch
+
+```go
+// Implements CPUManagerPolicy.
+type dynamicManager struct {}
+
+func (m *dynamicManager) Init(driver CPUDriver, topo CPUTopo) {
+  // TODO
+}
+
+func (m *dynamicManager) Add(c v1.Container, qos QoS) error {
+  // TODO
+}
+
+func (m *dynamicManager) Remove(c v1.Container, qos QoS) error {
+  // TODO
+}
+```
+
+##### Example pod specs and interpretation
+
+| Pod                                        | Interpretation                 |
+| ------------------------------------------ | ------------------------------ |
+|                                            |                                |
+|                                            |                                |
+
+## Operations and observability
+
+* Checkpointing assignments
+  * The CPU Manager must be able to pick up where it left off in case the
+    Kubelet restarts for any reason.
+* Read effective CPU assinments at runtime for alerting. This could be
+  satisfied by the checkpointing requirement.
+* Configuration
+  * How does the CPU Manager coexist with existing kube-reserved
+    settings?
+  * How does the CPU Manager coexist with related Linux kernel
+    configuration (e.g. `isolcpus`.) The operator may want to specify a
+    low-water-mark for the size of the shared cpuset. The operator may
```

Comment:
In the PoC branch we have the static policy unconditionally reserving a core for infra processes, with banning BE if the shared pool becomes empty as a TODO. However, containers in the Burstable class also might lack cpu requests (a single-container pod that requires only memory is in the Burstable class.)

@sjenning what do you think about replacing the notion of infra-reserved cores with a min size for the shared pool in the static policy? One thing I wouldn't want is to effectively disable a significant fraction of available threads on small systems (e.g. 4 or 8 cores.)